### ISSUE TITLE
specify samples by population name

### ISSUE THREAD

**Issue by petrelharp** [tskit-dev/tskit#1697]
It'd be nice to do e.g.
```
ts.samples("popA")
```
where `"popA"` is the *name* of a population, rather than the index.

**Comment by jeromekelleher** [comment]
I'm totally +1 for this. I think we can blur the line a bit here with the data/metadata distinction by saying it's data if we use it in the **C** library. Also, if we do decide we want to add a ``name`` field to the population at some point (which wouldn't be a terrible idea) we can easily imagine dealing with that here by saying the ``name`` field takes precedence, and if this is null we check for a metadata key called ``name``.

**Comment by benjeffery** [comment]
Yes, I'm happy with this, and with the idea that the C API defines what is metadata/data. Especially as we start to standardise metadata, I see it that there are three tiers:
 - data: used by C API and in tables
 - standard metadata: used by the Python API
 - user metadata: not used by tskit

**Comment by hyanwong** [comment]
Also not that you "can" do `ts.samples(population="A")` and it returns an empty array, which I think should be counted as a bug.

**Comment by benjeffery** [comment]
Some more up-to-date notes as this issue got posted as dupe in #2294.

This is very close to https://github.com/tskit-dev/tskit/issues/2239 where there is a proposal for the `TreeSequence` row-accessors to get similar functionality.
Personally I would prefer an explicit `ts.samples(name='BIG')` to `ts.samples('BIG')`. One option would be to make this the equivalent of #2239 which would be `ts.samples(metadata=("name", "BIG"))`

For the syntactically sweeter options `name` is promoted to "a metadata key used by the Python API" which means the standardization mentioned in https://github.com/tskit-dev/tskit/issues/1051 is relevant.

**Comment by hyanwong** [comment]
> Personally I would prefer an explicit `ts.samples(name='BIG')` to `ts.samples('BIG')`

If so, I suggest that `ts.samples('BIG')` should raise an error.

**Comment by hyanwong** [comment]
I just hit this "bug" again, when I tried to use `ts.samples(population='BIG')`. Note that In @benjeffery's suggestion, it's not clear that the 'BIG' is the name of a population (not, e.g an individual). It so happens that the first argument of `ts.samples` is `population`, but that seems a bit obscure to me. Perhaps

`ts.samples(population_name='big')`

or allow any argument to take a metadata dictionary:

```python
ts.samples(population={'name': 'big')  # if dict, assume we look in the metadata
```

**Comment by hyanwong** [comment]
Returning to this when looking through some teaching materials. What if we allowed the `population` parameter to `.samples()` to be either an integer or a dict. If a dict, the keys are taken to match against the metadata, as in https://github.com/tskit-dev/tskit/issues/2239.

That would mean a syntax like:

```python
ts.samples(0)
# or
ts.samples({'name':'EUR'})
```

That's flexible enough to then be able to do other metadata qualifications, but without having to make `name` a special kwarg etc.
